### ISSUE TITLE
Bug(planner): wrong output in hash join

### ISSUE THREAD

**Issue by xudong963** [datafuselabs/databend#5522]
The result is not correct in my case:
```
create table t1(a int, b int);
create table t2(a int, b int);
create table t3(a int, b int);

insert into t1 select number, number+1 from numbers(100000000);
insert into t2 select number, number+1 from numbers(20000000);
insert into t3 select number, number+1 from numbers(10000000);

select count(*) from t1 inner join t2 on t1.a=t2.a inner join t3 on t2.b=t3.b;

+----------+
| COUNT(*) |
+----------+
|        0 |
+----------+
1 row in set (6.45 sec)
```

Result should be:
```
┌──count()─┐
│ 10000000 │
└──────────┘
```

_Originally posted by @BohuTANG in https://github.com/datafuselabs/databend/issues/5497#issuecomment-1133793121_

**Comment by sundy-li** [comment]
```
fn visit_read_data_source(&mut self, plan: &ReadDataSourcePlan) -> Result<Pipeline> {
        // Bind plan partitions to context.
 self.ctx.try_set_partitions(plan.parts.clone())?;
```

I believe it's because we set the partition to the ctx. Since we read two different tables, the partition info are overridden.

cc @zhang2014 

```
:) select count(*) from (select to_int32(a) as a from t1 order by a limit 10) t1 inner join (select to_int32(number) as a from numbers(10)) t2 on t1.a=t2.a;
ERROR 1105 (HY000): Code: 1015, displayText = Cannot downcast from PartInfo to FusePartInfo. (while in processor thread 5).
```

**Comment by leiysky** [comment]
@sundy-li This should have been fixed by https://github.com/datafuselabs/databend/blob/eb07b393c769b152ef742c0b8f91032de5c98562/query/src/sql/exec/mod.rs#L191

**Comment by zhang2014** [comment]
This is strange, because t1 and t2 have same table engine(fuse). Maybe we should set `RUST_BACKTRACE=1` to get more info?

**Comment by zhang2014** [comment]
> @sundy-li This should have been fixed by

https://github.com/datafuselabs/databend/blob/eb07b393c769b152ef742c0b8f91032de5c98562/query/src/sql/exec/mod.rs#L319


We still use self.ctx in `build_physical_scan` CC: @leiysky

**Comment by leiysky** [comment]
@zhang2014 You are right, I will fix this in #5538 as well.

**Comment by leiysky** [comment]
I missed the related commit in #5538, thus I opened #5539 to fix this.
